zk_toolbox/crates/zk_inception/src/utils/ports.rs

```diff
@@ -0,0 +1,303 @@
+        self.ports.entry(port).or_default().push(info);
+    }
+
+    pub fn allocate_port(&mut self, range: Range<u16>) -> Result<u16> {
+        for port in range {
+            if !self.is_port_assigned(port) {
+                self.add_port_info(port, "Dynamically allocated".to_string());
+                return Ok(port);
+            }
+        }
+        anyhow::bail!("No available ports in the given range")
+    }
+
+    /// Allocates a set of ports based on given base ports, incrementing by offset if needed.
+    /// Tries base ports first, then base + offset, base + 2*offset, etc., until finding
+    /// available ports or exceeding the range. Returns allocated ports or an error if
+    /// no suitable ports are found.
+    pub fn allocate_ports(
+        &mut self,
+        base_ports: &[u16],
+        range: Range<u16>,
+        offset: u16,
+    ) -> Result<Vec<u16>> {
+        let mut i = 0;
+        loop {
+            let candidate_ports: Vec<u16> =
+                base_ports.iter().map(|&port| port + i * offset).collect();
+
+            // Check if all candidate ports are within the range
+            if candidate_ports.iter().any(|&port| !range.contains(&port)) {
+                anyhow::bail!(
+                    "No suitable ports found within the given range {:?}. Tried {} iterations with offset {}",
+                    range,
+                    i,
+                    offset
+                )
+            }
+
+            // Check if all candidate ports are available
+            if candidate_ports
+                .iter()
+                .all(|&port| !self.is_port_assigned(port))
+            {
+                // Allocate all ports
+                for &port in &candidate_ports {
+                    self.add_port_info(port, "Dynamically allocated".to_string());
+                }
+                return Ok(candidate_ports);
+            }
+            i += 1;
+        }
+    }
+}
+
+impl fmt::Display for EcosystemPorts {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut port_vec: Vec<_> = self.ports.iter().collect();
+        port_vec.sort_by_key(|&(key, _)| key);
+        for (port, port_infos) in port_vec {
+            for port_info in port_infos {
+                writeln!(f, "{} > {}", port_info, port)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+pub struct EcosystemPortsScanner {}
+
+impl EcosystemPortsScanner {
+    /// Scans the ecosystem directory for YAML files and extracts port information.
+    /// Specifically, it looks for keys ending with "port" and collects their values.
+    /// Note: Port information from Docker Compose files will not be picked up by this method.
+    pub fn scan(shell: &Shell) -> Result<EcosystemPorts> {
+        let _ = EcosystemConfig::from_file(shell)?; // Find and validate the ecosystem directory
+        let ecosystem_dir = shell.current_dir();
+
+        let skip_dirs: HashSet<&str> = SKIP_DIRS.iter().cloned().collect();
```

Comment:
Maybe I can use `ecosystem_config` as a parameter? Some methods inside it might be useful to verify the config folders we want to traverse.
Also can I rely that after `EcosystemConfig::from_file`, shell.current_dir() returns ecosystem root? Somehow, I feel an urge to implement `get_ecosystem_root()` inside `EcosystemConfig`.